src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php

```diff
@@ -125,7 +128,7 @@ public function preSubmit(FormEvent $event)
                 if (!$form->has($name)) {
                     $form->add($name, $this->type, array_replace(array(
                         'property_path' => '['.$name.']',
-                    ), $this->options));
+                    ), $this->getOptionsForEntry($value)));
```

Comment:
@HeahDude even with a refactoring, we will still need to add elements based on 2 different formats, as one case wants to add fields when knowing the model data and the other case when knowing the view data (and this cannot change with any refactoring, as it is part of the basic about how a collection type should work).

Btw, one of your previous comment assumes that the view data is a string. this assumption is wrong. you can perfectly have collections of objects (and then a compound form type as the entry type)